plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java

```diff
@@ -101,12 +106,35 @@ private InputStream openStream() throws IOException {
         }
     }
 
+    private long getStreamLength(final S3Object object) {
+        final ObjectMetadata metadata = object.getObjectMetadata();
+        try {
+            // Returns the content range of the object if response contains the Content-Range header.
+            final Long[] range = metadata.getContentRange();
+            if (range != null) {
+                assert range[1] >= range[0] : range[1] + " vs " + range[0];
+                assert range[0] == start + currentOffset :
+                    "Content-Range start value [" + range[0] + "] exceeds start [" + start + "] + current offset [" + currentOffset + ']';
+                assert range[1] == end : "Content-Range end value [" + range[1] + "] exceeds end [" + end + ']';
+                return range[1] - range[0] + 1L;
+            }
+            return metadata.getContentLength();
+        } catch (Exception e) {
+            assert false : e;
+            return Long.MAX_VALUE - 1L; // assume a large stream so that the underlying stream is aborted on closing, unless eof is reached
```

Comment:
Same here, maybe we should just use our own `end` and `start` offsets or use `metadata.getContentLength()` if we don't have an `end` instead of going through the indirection of the SDH header parsing here? That seems a lot more straight forward to me and doesn't require us to be scared of random exceptions from SDK misbehavior?

Also, then we could just make our life real easy. If `eof` is set to `true` or `start + currentOffset == currentStreamLastOffset` -> close, else `abort`. No need to even get the length from the metadata because any open ended stream of unknown length we'd read till EOF anyway?

Reply:
> No need to even get the length from the metadata because any open ended stream of unknown length we'd read till EOF anyway?

That's what I'm trying to avoid here; if we know the exact range or if we don't know it, the S3 endpoint should return the content length and we can use it to know if all bytes were really consumed. The exceptional case here should never happen and in this case we set an extra large `end` which should force anyway the stream to be aborted before closing.

Reply:
Fair point, I guess I still don't really like that we redundantly store the lengths and offsets here to some degree, but I also just noticed we we do the same in the GCS stream as well. I suppose this approach is the safest for now :) => let's go with it then.